src/librustc_trans/partitioning.rs

```diff
@@ -0,0 +1,376 @@
+            for &inlined_item in inlined_items {
+                follow_inlining(inlined_item, inlining_map, visited);
+            }
+        }
+    }
+}
+
+fn characteristic_def_id_of_trans_item<'tcx>(tcx: &TyCtxt<'tcx>,
+                                             trans_item: TransItem<'tcx>)
+                                             -> Option<DefId> {
+    match trans_item {
+        TransItem::Fn(instance) => {
+            // For now, just use the actual DefId of the item, with the
+            // consequence that, for example, methods are emitted in the
+            // module where the impl is located instead of the one where
+            // the self-type is located.
+            Some(instance.def)
+        }
+        TransItem::DropGlue(t) => characteristic_def_id_of_type(t),
+        TransItem::Static(node_id) => Some(tcx.map.local_def_id(node_id)),
+    }
+}
+
+fn compute_codegen_unit_name<'tcx>(tcx: &TyCtxt<'tcx>,
+                                   def_id: DefId,
+                                   volatile: bool)
+                                   -> InternedString {
+    // Unfortunately, TyCtxt::with_path() is a bit too clever for us here,
+    // because will make a difference between directly and indirectly
+    // referenced crates. We don't want that.
+    let mut mod_path = String::with_capacity(64);
+
+    if let Some(id) = tcx.map.as_local_node_id(def_id) {
+        // Local paths don't contain the crate name by themselves, so we
+        // prepend it.
+        mod_path.push_str(&tcx.crate_name[..]);
+        mod_path.push_str("-");
+        tcx.map.with_path(id, |p| push_up_to_first_non_mod(p, &mut mod_path));
+    } else {
+        push_up_to_first_non_mod(tcx.sess
+                                    .cstore
+                                    .extern_item_path(def_id)
+                                    .iter()
+                                    .cloned(),
+                                 &mut mod_path);
+    }
+
+    // Remove that last '-' separator
+    mod_path.pop();
+
+    if volatile {
+        mod_path.push_str(".volatile");
+    }
+
+    return token::intern_and_get_ident(&mod_path[..]);
+
+    // We only consider the path up to first non-module item.
+    fn push_up_to_first_non_mod<PI>(path: PI, output: &mut String)
+        where PI: Iterator<Item = hir_map::PathElem>
+    {
+        for path_elem in path {
+            match path_elem {
+                hir_map::PathMod(name) => {
+                    output.push_str(&name.as_str());
+                    output.push_str("-");
+                }
+                hir_map::PathName(_) => break,
+            }
+        }
+    }
+}
+
+// This function tries to find an associated DefId for a given type. This DefId
+// is used later to find the 'home' module for functions related to that type.
+//
+// There is a similar function in `ty::item_path` but it has slightly different
+// goals (human-readable def-paths) so we are not going to re-use it.
```

Comment:
Can we define formally define what a "characteristic def-id" is, independent of these two use cases? Otherwise, I'd rather keep the two functions separate as not to introduce an artificial dependency between them.